library/psa_crypto.c

```diff
@@ -3131,7 +3156,761 @@ psa_status_t psa_asymmetric_decrypt( mbedtls_svc_key_id_t key,
+
+    /* Update ops count with work done. */
+    operation->num_ops = psa_driver_wrapper_sign_hash_get_num_ops(operation);
+
+exit:
+
+    psa_wipe_output_buffer(signature, status, signature_size,
+            *signature_length);
+
+    if( status != PSA_OPERATION_INCOMPLETE )
+    {
+        if( status != PSA_SUCCESS )
+            operation->error_occurred = 1;
+
+        psa_sign_hash_abort_internal(operation);
+    }
+
+    return( status );
+}
+
+psa_status_t psa_sign_hash_abort(
+                            psa_sign_hash_interruptible_operation_t *operation )
+{
+    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
+
+    status = psa_sign_hash_abort_internal( operation );
+
+    /* We clear the number of ops done here, so that it is not cleared when
+     * the operation fails or succeeds, only on manual abort. */
+    operation->num_ops = 0;
+
+    /* Likewise, failure state. */
+    operation->error_occurred = 0;
+
+    return( status );
+}
+
+static psa_status_t psa_verify_hash_abort_internal(
+                        psa_verify_hash_interruptible_operation_t *operation )
+{
+    if( operation->id == 0 )
+    {
+        /* The object has (apparently) been initialized but it is not (yet)
+         * in use. It's ok to call abort on such an object, and there's
+         * nothing to do. */
+        return( PSA_SUCCESS );
+    }
+
+    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
+
+    status = psa_driver_wrapper_verify_hash_abort( operation );
+
+    operation->id = 0;
+
+    /* Do not clear either the error_occurred or num_ops elements here as they
+     * only want to be cleared by the application calling abort, not by abort
+     * being called at completion of an operation. */
+
+    return( status );
+}
+
+psa_status_t psa_verify_hash_start(
+                        psa_verify_hash_interruptible_operation_t *operation,
+                        mbedtls_svc_key_id_t key, psa_algorithm_t alg,
+                        const uint8_t *hash, size_t hash_length,
+                        const uint8_t *signature, size_t signature_length )
+{
+    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
+    psa_status_t unlock_status = PSA_ERROR_CORRUPTION_DETECTED;
+    psa_key_slot_t *slot;
+
+    /* Check that start has not been previously called, or operation has not
+     * previously errored. */
+    if( operation->id != 0 || operation->error_occurred )
+        return( PSA_ERROR_BAD_STATE );
+
+    status = psa_sign_verify_check_alg( 0, alg );
+    if( status != PSA_SUCCESS )
+    {
+        operation->error_occurred = 1;
+        return status;
+    }
+
+    status = psa_get_and_lock_key_slot_with_policy( key, &slot,
+                                                    PSA_KEY_USAGE_VERIFY_HASH,
+                                                    alg );
+
+    if( status != PSA_SUCCESS )
+    {
+        operation->error_occurred = 1;
+        return( status );
```

Comment:
[non-blocker] This is not tested according to lcov, and I'm not sure if it's easy to test (the equivalent for sign seems tested, though).

Reply:
#7113